archdoc/chap-altbounds.tex

```diff
@@ -35,4 +39,70 @@ \chapter{Potential revised bound encoding}
+This is similar to IEEE floating point choosing the exponent to \emph{normalize} the mantissa.
+This trick allows us to store only 8-bits of $T$ and reconstruct the top bit using the same technique as CHERI Concentrate\cite{Woodruff2019}, as explained below.
+Using the bit saved by this we can expand the exponent to 5 bits, meaning we can represent all exponents between 0 and 24 with some to spare.
+We can use one of the spare exponents, 31, to encode the \emph{subnormal} case for lengths less that 256 (with $L[8] = 0)$.
+Lastly, so that the all-zero capability encodes zero length we store the exponent bitwise inverted.
+
+Putting these things together the encoding is revised to that shown in \cref{fig:implicitTformat}.
+The only difference from \cref{fig:capformat} is that E is expanded to five bits by taking a bit from T.
+
+\begin{figure}[h]
+\begin{bytefield}[bitwidth=\linewidth/32]{32}
+    \bitheader[endianness=big]{0,8,9,16,17,21,22,24,25,31} \\
+    \bitbox{1}{R} & \bitbox{6}{$p$'6} & \bitbox{3}{otype'3} & \bitbox{4}{E'5} & \bitbox{9}{T'8} & \bitbox{9}{B'9} \\
+    \bitbox[lrb]{32}{$a$'32}
+\end{bytefield}
+\caption{\label{fig:implicitTformat}Capability encoding with implicit T[8].}
+\end{figure}
+
+To decode this we decode the exponent, $e$:
+
+\begin{center}
+\begin{tabular}{r c l}
+$e$ &=& $ \begin{cases}
+              0,& \text{if } E = 0 \\
+              \lnot E,& \text{otherwise (bitwise inversion)} \\
+\end{cases} $ \\
+\end{tabular}
+\end{center}
+
+And the top bit of the 9-bit length $L$:
+
+\begin{center}
+\begin{tabular}{r c l}
+$L[8]$ &=& $ \begin{cases}
+              0,& \text{if } E = 0 \\
+              1,& \text{otherwise} \\
+\end{cases} $ \\
+\end{tabular}
+\end{center}
+
+Note that this means there are two encodings with $e=0$: $E=0$ for lengths in the range $0 \dots 255$ (subnormal), and $E=31$ for lengths $256 \dots 511$ (normal).
+
+Now, since $T = B + L$ we can recover $T[8]$ using $T[8] = B[8] \oplus L[8] \oplus C$ where $C$ is a potential carry-in from the overflow of $B[7..0] + L[7..0]$.
+The encoding stores $T[7..0]$, not $L[7..0]$, but we can infer the carry in by observing that if $T[7..0] \lt B[7..0]$ an overflow must have occurred. Hence,
+
+\begin{center}
+\begin{tabular}{r c l}
+$C$ &=& $ \begin{cases}
+              1,& \text{if } T[7..0] \lt B[7..0] \\
+              0,& \text{otherwise} \\
+\end{cases} $ \\
+\end{tabular}
+\end{center}
+
+Bounds decoding can then continue exactly as in \cref{sec:bounds} using $a$, $e$, $B$ and the reconstituted $T$.
+
+The advantage of this encoding is that it eliminates the gap in encodings between $e=14$ and $e=24$, enabling much better bounds precision for lengths greater than about 8 MiB (see \cref{tab:caplen}) while using no extra bits.
+In terms of hardware the bounds decoding is only slightly complicated by the need to compute T[8].
```

Comment:
Yes, and this bit can likely be added to the partially decompressed capability correction bits we already have in CHERIoT Ibex.